File: src/converter.cpp
#include <emscripten/bind.h>
#include <emscripten/val.h>

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <unordered_set>
#include <vector>

using emscripten::val;

namespace {
inline double luma(double r, double g, double b) {
  return 0.299 * r + 0.587 * g + 0.114 * b;
}
inline uint8_t toByte(double v) {
  return v <= 0 ? 0 : v >= 255 ? 255 : static_cast<uint8_t>(std::lround(v));
}
// Snap a channel to one of the 16 .seurat levels (4-bit color): v -> round(v/17)*17.
inline uint8_t quant4(uint8_t v) {
  return static_cast<uint8_t>(std::lround(v / 17.0) * 17);
}
inline uint64_t key(int x, int y) {
  return (static_cast<uint64_t>(static_cast<uint32_t>(x)) << 32) |
         static_cast<uint32_t>(y);
}
}  // namespace

class DotConverter {
 public:
  void setContrast(double percent) { contrast_ = percent / 100.0; }
  void setExposure(double exposure) { exposure_ = exposure; }
  void setWhiteThreshold(double threshold) { whiteThreshold_ = threshold; }

  void clearMask() { masked_.clear(); }
  void addMaskPixel(int x, int y) { masked_.insert(key(x, y)); }

  int gridHeight(int srcWidth, int srcHeight, int gridWidth) const {
    if (srcWidth <= 0 || gridWidth <= 0) return 0;
    int h = static_cast<int>(std::floor(double(gridWidth) * srcHeight / srcWidth));
    return std::max(h, 1);
  }

  val convertFrame(val rgba, int width, int height) {
    const std::vector<uint8_t> src =
        emscripten::convertJSArrayToNumberVector<uint8_t>(rgba);
    const size_t n = size_t(width) * height;

    rgb_.resize(n * 3);
    for (size_t i = 0; i < n; ++i) {
      double r = src[i * 4 + 0] + exposure_;
      double g = src[i * 4 + 1] + exposure_;
      double b = src[i * 4 + 2] + exposure_;
      rgb_[i * 3 + 0] = toByte((r - 128) * contrast_ + 128);
      rgb_[i * 3 + 1] = toByte((g - 128) * contrast_ + 128);
      rgb_[i * 3 + 2] = toByte((b - 128) * contrast_ + 128);
    }

    double minL = 255, maxL = 0;
    for (size_t i = 0; i < n; ++i) {
      double l = luma(rgb_[i * 3], rgb_[i * 3 + 1], rgb_[i * 3 + 2]);
      minL = std::min(minL, l);
      maxL = std::max(maxL, l);
    }
    double range = maxL - minL;
    if (range == 0) range = 1;

    out_.resize(n * 3);
    for (size_t i = 0; i < n; ++i) {
      double r = rgb_[i * 3], g = rgb_[i * 3 + 1], b = rgb_[i * 3 + 2];
      double l = luma(r, g, b);
      double f = l > 0 ? ((l - minL) / range) * 255 / l : 1;
      uint8_t R = toByte(r * f), G = toByte(g * f), B = toByte(b * f);

      int x = int(i % width), y = int(i / width);
      bool empty = masked_.count(key(x, y)) || luma(R, G, B) >= whiteThreshold_;
      // Quantize to 4-bit color so the core output matches the .seurat format.
      out_[i * 3 + 0] = empty ? 255 : quant4(R);
      out_[i * 3 + 1] = empty ? 255 : quant4(G);
      out_[i * 3 + 2] = empty ? 255 : quant4(B);
    }
    return val(emscripten::typed_memory_view(out_.size(), out_.data()));
  }

 private:
  double contrast_ = 1.0;
  double exposure_ = 0.0;
  double whiteThreshold_ = 240;
  std::unordered_set<uint64_t> masked_;
  std::vector<uint8_t> rgb_;
  std::vector<uint8_t> out_;
};

EMSCRIPTEN_BINDINGS(seurat) {
  emscripten::class_<DotConverter>("DotConverter")
      .constructor<>()
      .function("setContrast", &DotConverter::setContrast)
      .function("setExposure", &DotConverter::setExposure)
      .function("setWhiteThreshold", &DotConverter::setWhiteThreshold)
      .function("clearMask", &DotConverter::clearMask)
      .function("addMaskPixel", &DotConverter::addMaskPixel)
      .function("gridHeight", &DotConverter::gridHeight)
      .function("convertFrame", &DotConverter::convertFrame);
}

File: web/src/lib/converter.ts
// Seurat dot converter — browser glue around the WebAssembly core.
// Decodes a video, downscales each frame to the dot grid, runs it through the
// WASM DotConverter, and packs the result into the .seurat format the player reads.

import createConverter from '../wasm/converter.js';
import { encodeSeurat } from './seurat.ts';

export interface ConvertOptions {
  fps?: number;
  gridWidth?: number;
  skipStartFrames?: number;
  skipEndFrames?: number;
  cropTop?: number;
}

export interface DotResult {
  rgbFrames: Uint8Array[];
  fps: number;
  width: number;
  height: number;
  frameCount: number;
  duration: number;
}

export interface Progress {
  stage: 'extracting' | 'converting' | 'complete';
  current?: number;
  total?: number;
  percent: number;
}

let modulePromise: Promise<any> | null = null;
const loadModule = () => (modulePromise ??= createConverter());

export class SeuratDotConverter {
  whiteThreshold: number;
  contrast: number;
  exposure: number;
  maskedPixels: Set<string>;
  onProgress: (p: Progress) => void;
  private conv: any = null;

  constructor(options: Partial<Pick<SeuratDotConverter, 'whiteThreshold' | 'contrast' | 'exposure' | 'maskedPixels' | 'onProgress'>> = {}) {
    this.whiteThreshold = options.whiteThreshold ?? 240;
    this.contrast = options.contrast ?? 100;
    this.exposure = options.exposure ?? 0;
    this.maskedPixels = options.maskedPixels ?? new Set();
    this.onProgress = options.onProgress ?? (() => {});
  }

  async ready() {
    if (!this.conv) {
      const Module = await loadModule();
      this.conv = new Module.DotConverter();
    }
    this.applySettings();
    return this.conv;
  }

  update(options: Partial<Pick<SeuratDotConverter, 'whiteThreshold' | 'contrast' | 'exposure' | 'maskedPixels'>> = {}) {
    Object.assign(this, options);
    if (this.conv) this.applySettings();
  }

  gridHeight(srcWidth: number, srcHeight: number, gridWidth: number): number {
    return this.conv.gridHeight(srcWidth, srcHeight, gridWidth);
  }

  // One already-downscaled RGBA grid frame -> packed RGB frame (for live preview).
  convertGridFrame(rgba: Uint8ClampedArray, width: number, height: number): Uint8Array {
    return this.conv.convertFrame(rgba, width, height).slice();
  }

  async convertToDots(videoFile: File, options: ConvertOptions = {}): Promise<DotResult> {
    const { fps = 10, gridWidth = 300, skipStartFrames = 0, skipEndFrames = 0, cropTop = 0 } = options;
    await this.ready();

    const video = document.createElement('video');
    video.muted = true;
    video.playsInline = true;

    try {
      await this.loadMetadata(video, videoFile);
      const srcW = video.videoWidth;
      const srcH = Math.max(1, video.videoHeight - cropTop);
      const width = gridWidth;
      const height = this.gridHeight(srcW, srcH, gridWidth);

      const grid = document.createElement('canvas');
      grid.width = width;
      grid.height = height;
      const ctx = grid.getContext('2d', { willReadFrequently: true })!;

      const rgbFrames: Uint8Array[] = [];
      await this.eachFrame(video, fps, skipStartFrames, skipEndFrames, (index, total) => {
        ctx.drawImage(video, 0, cropTop, srcW, srcH, 0, 0, width, height);
        const { data } = ctx.getImageData(0, 0, width, height);
        rgbFrames.push(this.convertGridFrame(data, width, height));
        this.onProgress({ stage: 'converting', current: index + 1, total, percent: Math.round(((index + 1) / total) * 100) });
      });

      this.onProgress({ stage: 'complete', percent: 100 });
      return { rgbFrames, fps, width, height, frameCount: rgbFrames.length, duration: rgbFrames.length / fps };
    } finally {
      video.removeAttribute('src');
      video.load();
    }
  }

  // Pack frames into a single .seurat file (additive delta, 4-bit color).
  async toSeurat(result: DotResult): Promise<Blob> {
    return encodeSeurat(
      { width: result.width, height: result.height, fps: result.fps, frameCount: result.frameCount },
      result.rgbFrames,
    );
  }

  dispose() {
    this.conv?.delete();
    this.conv = null;
  }

  private applySettings() {
    const c = this.conv;
    c.setContrast(this.contrast);
    c.setExposure(this.exposure);
    c.setWhiteThreshold(this.whiteThreshold);
    c.clearMask();
    for (const k of this.maskedPixels) {
      const [x, y] = k.split(',').map(Number);
      c.addMaskPixel(x, y);
    }
  }

  private loadMetadata(video: HTMLVideoElement, file: File): Promise<void> {
    return new Promise((resolve, reject) => {
      video.onloadedmetadata = () => resolve();
      video.onerror = () => reject(new Error('Failed to load video'));
      video.src = URL.createObjectURL(file);
    });
  }

  private eachFrame(
    video: HTMLVideoElement,
    fps: number,
    skipStart: number,
    skipEnd: number,
    onFrame: (index: number, total: number) => void,
  ): Promise<number> {
    return new Promise((resolve, reject) => {
      const total = Math.max(0, Math.floor(video.duration * fps) - skipStart - skipEnd);
      let i = 0;
      const seek = () => {
        if (i >= total) return resolve(total);
        video.currentTime = (skipStart + i) * (1 / fps);
      };
      video.onseeked = () => {
        onFrame(i, total);
        i++;
        this.onProgress({ stage: 'extracting', current: i, total, percent: Math.round((i / total) * 50) });
        setTimeout(seek, 0);
      };
      video.onerror = () => reject(new Error('Failed to read frame'));
      seek();
    });
  }
}
